Next failure here:
```
ctest -R PyQgsVirtualLayerProvider --output-on-failure
...
FAIL: testExtentSubsetString (__main__.TestQgsVirtualLayerProvider)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/src/qgis/qgis/src/rectangle-null/tests/src/python/providertestbase.py", line 462, in testExtentSubsetString
    self.assertEqual(provider_extent, QgsRectangle())
AssertionError: <QgsRectangle: 0 0, 0 0> != <QgsRectangle: 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368, -179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368 -179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368>
```

--- a/src/core/geometry/qgsrectangle.cpp
+++ b/src/core/geometry/qgsrectangle.cpp
@@ -146,8 +146,8 @@ QString QgsRectangle::toString( int precision ) const
     }
   }
 
-  if ( isEmpty() )
-    rep = QStringLiteral( "Empty" );
+  if ( isNull() )
+    rep = QStringLiteral( "Null" );
   else
     rep = QStringLiteral( "%1,%2 : %3,%4" )
           .arg( mXmin, 0, 'f', precision )

--- a/src/core/geometry/qgsrectangle.h
+++ b/src/core/geometry/qgsrectangle.h
@@ -516,7 +516,7 @@ class CORE_EXPORT QgsRectangle
      */
     bool isEmpty() const
     {
-      return mXmax < mXmin || mYmax < mYmin || qgsDoubleNear( mXmax, mXmin ) || qgsDoubleNear( mYmax, mYmin );
+      return isNull() || mXmax <= mXmin || mYmax <= mYmin || qgsDoubleNear( mXmax, mXmin ) || qgsDoubleNear( mYmax, mYmin );
     }
 
     /**
@@ -533,7 +533,6 @@ class CORE_EXPORT QgsRectangle
       // rectangle created QgsRectangle() or with rect.setNull() or
       // otherwise having NaN ordinates
       return ( std::isnan( mXmin )  && std::isnan( mXmax ) && std::isnan( mYmin ) && std::isnan( mYmax ) ) ||
-             ( qgsDoubleNear( mXmin, 0.0 ) && qgsDoubleNear( mXmax, 0.0 ) && qgsDoubleNear( mYmin, 0.0 ) && qgsDoubleNear( mYmax, 0.0 ) ) ||
              ( qgsDoubleNear( mXmin, std::numeric_limits<double>::max() ) && qgsDoubleNear( mYmin, std::numeric_limits<double>::max() ) &&
                qgsDoubleNear( mXmax, -std::numeric_limits<double>::max() ) && qgsDoubleNear( mYmax, -std::numeric_limits<double>::max() ) );
     }
@@ -665,10 +664,10 @@ class CORE_EXPORT QgsRectangle
 
   private:
 
-    double mXmin = 0.0;
-    double mYmin = 0.0;
-    double mXmax = 0.0;
-    double mYmax = 0.0;
+    double mXmin = std::numeric_limits<double>::max();
+    double mYmin = std::numeric_limits<double>::max();
+    double mXmax = -std::numeric_limits<double>::max();
+    double mYmax = -std::numeric_limits<double>::max();
 
 };
 

--- a/tests/src/core/geometry/testqgsrectangle.cpp
+++ b/tests/src/core/geometry/testqgsrectangle.cpp
@@ -72,10 +72,12 @@ void TestQgsRectangle::isNull()
   QVERIFY( QgsRectangle().isNull() );
   QVERIFY( QgsRectangle( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN(),
                          std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ).isNull() );
-  QVERIFY( QgsRectangle( 0.0, 0.0, 0.0, 0.0 ).isNull() );
+  QVERIFY( !QgsRectangle( 0.0, 0.0, 0.0, 0.0 ).isNull() );
+  QVERIFY( !QgsRectangle( 1.0, 1.0, 1.0, 1.0 ).isNull() );
   QVERIFY( !QgsRectangle( std::numeric_limits<double>::max(), -std::numeric_limits<double>::max(), std::numeric_limits<double>::max(), -std::numeric_limits<double>::max() ).isNull() );
   QVERIFY( !QgsRectangle( 1, 2, 2, 1 ).isNull() );
 }
+
 
 void TestQgsRectangle::fromWkt()
 {

--- a/tests/src/python/providertestbase.py
+++ b/tests/src/python/providertestbase.py
@@ -249,6 +249,10 @@ class ProviderTestCase(FeatureSourceTestCase):
             The WFS provider might not always be able to have that guarantee. """
         return True
 
+    def referenceExtent(self):
+        """ Extent of the reference dataset """
+        return QgsRectangle(-71.123, 66.33, -65.32, 78.3)
+
     def getSubsetString(self):
         """Individual providers may need to override this depending on their subset string formats"""
         return '"cnt" > 100 and "cnt" < 410'
@@ -256,6 +260,10 @@ class ProviderTestCase(FeatureSourceTestCase):
     def getSubsetString2(self):
         """Individual providers may need to override this depending on their subset string formats"""
         return '"cnt" > 100 and "cnt" < 400'
+
+    def referenceSubsetString3Extent(self):
+        """ Extent of the data selected by subset string 3 """
+        return QgsRectangle(-68.2, 70.8, -68.2, 70.8)
 
     def getSubsetString3(self):
         """Individual providers may need to override this depending on their subset string formats"""
@@ -422,13 +430,12 @@ class ProviderTestCase(FeatureSourceTestCase):
             self.assertEqual(max_value, 300)
 
     def testExtent(self):
-        reference = QgsGeometry.fromRect(
-            QgsRectangle(-71.123, 66.33, -65.32, 78.3))
+        reference_extent = self.referenceExtent()
         provider_extent = self.source.extent()
-        self.assertAlmostEqual(provider_extent.xMinimum(), -71.123, 3)
-        self.assertAlmostEqual(provider_extent.xMaximum(), -65.32, 3)
-        self.assertAlmostEqual(provider_extent.yMinimum(), 66.33, 3)
-        self.assertAlmostEqual(provider_extent.yMaximum(), 78.3, 3)
+        self.assertAlmostEqual(provider_extent.xMinimum(), reference_extent.xMinimum())
+        self.assertAlmostEqual(provider_extent.xMaximum(), reference_extent.xMaximum())
+        self.assertAlmostEqual(provider_extent.yMinimum(), reference_extent.yMinimum())
+        self.assertAlmostEqual(provider_extent.yMaximum(), reference_extent.yMaximum())
 
     def testExtentSubsetString(self):
         if self.source.supportsSubsetString():
@@ -437,12 +444,13 @@ class ProviderTestCase(FeatureSourceTestCase):
             self.source.setSubsetString(subset)
             count = self.source.featureCount()
             provider_extent = self.source.extent()
+            subset_extent = self.referenceSubsetString3Extent()
             self.source.setSubsetString(None)
             self.assertEqual(count, 1)
-            self.assertAlmostEqual(provider_extent.xMinimum(), -68.2, 3)
-            self.assertAlmostEqual(provider_extent.xMaximum(), -68.2, 3)
-            self.assertAlmostEqual(provider_extent.yMinimum(), 70.8, 3)
-            self.assertAlmostEqual(provider_extent.yMaximum(), 70.8, 3)
+            self.assertAlmostEqual(provider_extent.xMinimum(), subset_extent.xMinimum(), 3)
+            self.assertAlmostEqual(provider_extent.xMaximum(), subset_extent.xMaximum(), 3)
+            self.assertAlmostEqual(provider_extent.yMinimum(), subset_extent.yMinimum(), 3)
+            self.assertAlmostEqual(provider_extent.yMaximum(), subset_extent.yMaximum(), 3)
 
             # with no points
             subset = self.getSubsetStringNoMatching()

--- a/tests/src/python/providertestbase.py
+++ b/tests/src/python/providertestbase.py
@@ -451,7 +451,7 @@ class ProviderTestCase(FeatureSourceTestCase):
             provider_extent = self.source.extent()
             self.source.setSubsetString(None)
             self.assertEqual(count, 0)
-            self.assertTrue(provider_extent.isNull())
+            self.assertEqual(provider_extent, QgsRectangle())
             self.assertEqual(self.source.featureCount(), 5)
 
     def testUnique(self):

--- a/tests/src/python/test_qgsrectangle.py
+++ b/tests/src/python/test_qgsrectangle.py
@@ -125,7 +125,7 @@ class TestQgsRectangle(QgisTestCase):
 
     def testToString(self):
         """Test the different string representations"""
-        self.assertEqual(QgsRectangle().toString(), 'Empty')
+        self.assertEqual(QgsRectangle().toString(), 'Null')
         rect = QgsRectangle(0, 0.1, 0.2, 0.3)
         self.assertEqual(rect.toString(), '0.0000000000000000,0.1000000000000000 : 0.2000000000000000,0.3000000000000000')
 

